File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.dart
class Solution {
  int strStr(String haystack, String needle) {
    int n = haystack.length, m = needle.length;
    if (m == 0) return 0;
    List<int> lps = List.filled(m, 0);
    for (int i = 1, len = 0; i < m;) {
      if (needle[i] == needle[len]) {
        lps[i++] = ++len;
      } else if (len > 0) {
        len = lps[len - 1];
      } else {
        lps[i++] = 0;
      }
    }
    for (int i = 0, j = 0; i < n;) {
      if (haystack[i] == needle[j]) {
        i++; j++;
        if (j == m) return i - m;
      } else if (j > 0) {
        j = lps[j - 1];
      } else {
        i++;
      }
    }
    return -1;
  }
}

File: sol/solution/0001-0100/0028. Find the Index of the First Occurrence in a String/Solution.rs
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let n = haystack.len();
        let m = needle.len();
        if m == 0 { return 0; }
        let h: Vec<char> = haystack.chars().collect();
        let nd: Vec<char> = needle.chars().collect();
        let mut lps = vec![0; m];
        let (mut len, mut i) = (0, 1);
        while i < m {
            if nd[i] == nd[len] {
                len += 1;
                lps[i] = len;
                i += 1;
            } else if len > 0 {
                len = lps[len - 1];
            } else {
                lps[i] = 0;
                i += 1;
            }
        }
        let (mut j, mut idx) = (0, 0);
        while idx < n {
            if h[idx] == nd[j] {
                idx += 1; j += 1;
                if j == m { return (idx - m) as i32; }
            } else if j > 0 {
                j = lps[j - 1];
            } else {
                idx += 1;
            }
        }
        -1
    }
}